Issue 262 publint audit

Sukuna0007Abhi opened the pull request:
<!--
Thank you for contributing to Caravan! Please provide the following details to help us review your pull request efficiently.
-->

Fixes #262

Issue proof: 
![IMG-20251204-WA0004](https://github.com/user-attachments/assets/1ccc7941-dad6-4982-ad59-ab68de83f5de)

![IMG-20251204-WA0003](https://github.com/user-attachments/assets/4d224e14-b2ea-4033-8db0-7699dd6439ec)

![IMG-20251204-WA0001](https://github.com/user-attachments/assets/7074d8df-0e77-483e-a792-367e615c7737)

![IMG-20251204-WA0005](https://github.com/user-attachments/assets/cd72fce3-5fb0-4f4c-8e2e-df5bc5084bc5)


This PR addresses all publint warnings and suggestions for the published packages by implementing proper module type declarations and export conditions.

- ****I don't added any documentation related to it**...**

### Changes Made

**All packages (@caravan/bitcoin, @caravan/psbt, @caravan/wallets, @caravan/multisig):**
- Added `"type": "commonjs"` field for proper Node.js module detection
- Updated `exports` field with separate type conditions for ESM and CommonJS
- Added `.d.mts` generation for proper ESM TypeScript support
- Included `.d.mts` files in the published package files array
- Updated build scripts to automatically generate `.d.mts` from `.d.ts`
- Fixed repository URLs to use proper `git+https://` format with `directory` field for monorepo packages

**@caravan/bip32:**
- Since this package uses `"type": "module"`, implemented `.d.cts` generation for CommonJS types
- Updated exports to use `.d.cts` for require condition and `.d.mts` for import condition
- Modified generation script to use ESM syntax (import/export) instead of CommonJS

**@caravan/clients:**
- Added missing `exports` field (previously only had `module` field which Node.js ignores)
- Configured proper ESM/CommonJS export conditions with corresponding type declarations

### Testing

All packages now pass publint validation with zero warnings or errors. Verified locally using:

```
npx publint packages/caravan-bitcoin
npx publint packages/caravan-psbt
npx publint packages/caravan-wallets
npx publint packages/bip32
npx publint packages/multisig
npx publint packages/caravan-clients
```

### Proofs: 
![IMG-20251204-WA0007](https://github.com/user-attachments/assets/dd16d235-83ca-4320-ac4c-cec140e30bc5)
![IMG-20251204-WA0008](https://github.com/user-attachments/assets/5a09f662-cc39-4a1c-915d-630315617126)
![IMG-20251204-WA0009](https://github.com/user-attachments/assets/3ca0b730-fa43-48cd-aaef-6be28e3d021d)


### Breaking Changes

The addition of `exports` field to @caravan/clients is technically a breaking change as it changes how the package can be imported, though it enables proper ESM support that was previously not working correctly.

All other changes are backwards compatible improvements to package metadata and type declarations.

Ready for review @bucko13 @Legend101Zz

Sukuna0007Abhi committed:
fix: audit and improve package definitions using publint

Fixes #262

Conducted comprehensive publint audit and implemented all fixes:
- Added type: commonjs field to 5 packages for proper Node.js module detection
- Fixed repository URLs to proper git+https format with directory field
- Added .d.mts generation scripts for proper ESM type support

Packages updated: @caravan/bitcoin, @caravan/psbt, @caravan/wallets,
@caravan/clients, @caravan/multisig, @caravan/bip32

All publint suggestions and warnings now resolved.

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: use ESM syntax in bip32 generate-mts-dts script

bip32 package has type: module so the script needs ESM imports

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: generate .d.cts for bip32 CommonJS types

Since bip32 has type: module, .d.ts files are treated as ESM.
For proper CommonJS type support, we need .d.cts extension.

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: add exports field to caravan-clients package

pkg.module is not read by Node.js without pkg.exports defined.
Added proper exports field with ESM/CJS conditions and .d.mts types.

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
Use tsup build system to generate declaration files

- Replace manual scripts with tsup onSuccess hooks
- Add onSuccess hooks to all tsup.config files to generate .d.mts and .d.cts
- Remove manual generate-mts-dts.js scripts from all packages
- Clean up package.json build commands

This addresses the maintainer feedback to use the build system (tsup)
instead of manual scripts for generating module type declaration files.

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Legend101Zz commented:
@Sukuna0007Abhi thanks for the PR and sorry for the delay in reviews ... I'll try to find some time today to take a look at this ...

Legend101Zz commented on packages/caravan-bitcoin/package.json:
```diff
@@ -2,24 +2,32 @@
   "repository": {
     "type": "git",
-    "url": "https://github.com/caravan-bitcoin/caravan/packages/caravan-bitcoin"
+    "url": "git+https://github.com/caravan-bitcoin/caravan.git",
```
why was this changed ?

Legend101Zz commented on packages/caravan-bitcoin/tsup.config.ts:
```diff
@@ -1,11 +1,23 @@
     provideSelf(),
     provideNavigator(),
   ],
+  onSuccess: async () => {
```
again remove the file copying and let tsup generate both

Legend101Zz reviewed (changes requested):
Thanks for the PR ... a couple of requests from my end , mainly regarding the opying index.d.ts → index.d.mts and index.d.cts stuff

Sukuna0007Abhi committed:
build: rely on tsup for dual outputs

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
chore: emit bip32 per-format dts

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: bip 32 emit per-format dts                                                                                  Signed-off-by:Sukuna0007Abhi <appsonl310@gmail.com>

Sukuna0007Abhi committed:
fix: use stable dts for bip32

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
feat: emit dual dts for bip32

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
chore: fix bip32 dts emit hook import

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
chore: build bip32 before pack

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: point bip32 types to generated d.ts

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
chore(bip32): export format-specific dts

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
chore(bip32): ensure dual dts on build

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
chore(psbt): align types mapping with single dts

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
Buffer to Uint8Array                             Signed-off-by: Sukuna0007Abhi and  <appsonly310@gmail.com>

Legend101Zz reviewed (changes requested):
@bucko13 I guess this is the reason why the ci was failing

Sukuna0007Abhi committed:
fix: use postbuild script instead of onSuccess hook

- bip32: Move dual dts generation to postbuild to fix timing issues
- Remove onSuccess hook from tsup config (causes race condition)
- Simplify other packages to single .d.ts (type: commonjs)
- Fix psbt Buffer type error with type assertion

Addresses mentor feedback on PR #417

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): type assert buffer slice\n\nSigned-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): satisfy BufferReader type signature\n\nSigned-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): ensure serialize Buffer typing\n\nSigned-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): use Uint8Array view for BufferReader\n\nSigned-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): silence BufferReader type with any cast

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): cast all BufferReader instantiations

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): cast render result and encoding type

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): remove explicit Buffer type annotation

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): use any cast for toString call

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(psbt): cast Buffer.from argument

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(transactions): cast Buffer methods for dts build

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(transactions): cast Buffer.reverse in transactionAnalyzer

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(transactions): cast Buffer.from input.hash

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(transactions): cast Buffer.from bip32 derivation fields

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(coordinator): cast Buffer operations in Hermit importer

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix(coordinator): cast Buffer to ArrayBuffer for parseBinaryPath

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: add .d.mts declarations for publint compliance

- Add .d.mts to files array and exports for ESM type resolution
- Update exports structure to nest types under import/require conditions
- Add postbuild scripts to generate .d.mts from .d.ts
- Fix repository URLs and add directory field for monorepo packages
- Fixes publint warnings for @caravan/bitcoin, @caravan/wallets, @caravan/multisig, @caravan/psbt

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
chore: add emit-mts scripts for postbuild hooks

- Add scripts/emit-mts.js for packages that need .d.mts generation
- Required by postbuild hooks in package.json

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: convert emit-mts scripts to CommonJS syntax

- Change import statements to require() for CommonJS compatibility
- Fixes build errors in packages with "type": "commonjs"

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi commented:
Yes, sorry sir @bucko13 , for late response actually, what the issue is happening because of this warning:
Which I also wanted to fix, but can't fix it yet...
![IMG-20260109-WA0000](https://github.com/user-attachments/assets/26b4c0da-c87c-487b-bb6b-a2f584f29a93)

Why I created the scripts actually because:
Publint requires .d.mts files for ESM imports (as my screenshot shows)
tsup by default only generates .d.ts files, I tried with tsup.config.js using `experimentalDts` also but it's still not working for generating...
So that's why I added postbuild scripts to copy .d.ts → .d.mts

Still, though I am still trying and testing everyday to find more robust approach, which will align with your idea without creating custom scripts..

I will ping you sir, when I completed it...

Legend101Zz commented:
> If we are ok with this warning then I think the job is already done I will just reset these commited scripts.. what your thoughts about it pls help @bucko13 @Legend101Zz

@Sukuna0007Abhi thanks a ton for sticking with this issue 🙏
For now, could you sync your branch with `main`? I’m seeing that *“This branch is out-of-date with the base branch”*.

Once that’s done, I’ll pull it locally, run everything, and test it out. Overall, it looks good from my side — and thanks again for staying on it. These build toolchain issues are definitely tricky to deal with 😄

Let me test this and get back to you tomorrow...

Sukuna0007Abhi committed:
refactor: use tsup experimentalDts for .d.mts generation

- Replace custom emit-mts.js scripts with tsup built-in functionality
- Add tsup.config files with experimentalDts.resolve option
- Remove postbuild scripts from package.json
- Delete custom emit-mts.js scripts

This uses tsup's native capabilities instead of maintaining custom scripts.

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
Merge branch 'main' into issue-262-publint-audit

Sukuna0007Abhi commented:
## What changes I did at the end after several attempts:
- Added tsup.config.ts files to package (bitcoin, clients, wallets, multisig, psbt)
- Although transactions package is private I still put some code on it also, pls check if you think no need, then I will remove it
- Removed all custom scripts as sir @bucko13 @Legend101Zz said.

## Note: `@caravan/bip32` needs to use custom scripts because it:
- It generates both .d.mts (ESM) and .d.cts(CommonJs) declaration files
- Performs a special transformation something like on `.d.cts: ` converts `export deafault` syntax to `export =` for proper CommonJs typing
- And this transformation isn't currently handled by tsup's experimentalDts

So, that's why this custom scripts will help for this dual module setup..

Sukuna0007Abhi commented:
Ready for review @bucko13 @Legend101Zz  & I updated the branch...

Legend101Zz reviewed (approved):
@Sukuna0007Abhi thanks for the changes they look good to me now , just I feel that regex for transforming default exports is too simplistic and will break on complex type declarations , but I'll let @bucko13 have the final call on this ...

Thanks a ton for the PR and your patience for sticking out with this :)

Sukuna0007Abhi commented:
Pls review @bucko13 , & thanks for the review @Legend101Zz , maybe hoping this will be merged via this week...

Sukuna0007Abhi committed:
Merge branch 'main' into issue-262-publint-audit

Sukuna0007Abhi committed:
refactor: improve default export regex in bip32 dts script

Simplified comments and improved regex to extract declaration name properly.

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
Merge branch 'main' into issue-262-publint-audit

Sukuna0007Abhi commented:
> > So if there were no commonjs export for BIP32 then the custom transform wouldn't be necessary? It seems like the ecosystem is moving away from needing commonjs so if we set the right minimum engine in the package.json and just avoid that entirely than that would be fine too.
> 
> Sorry my net got bumped,
> 
> Yeah, fully correct , it's actually perfect, we can do this

So, should I revert the changes I did for bip32 & make sure package.json has the right minimum node version @bucko13

Sukuna0007Abhi committed:
refactor: make bip32 ESM-only, remove CommonJS and custom transform script

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: align bip32 config with other packages pattern

- Add explicit outExtension in tsup.config for consistency
- Add format flag in build script to match other packages
- Keep ESM-only structure as intended

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
fix: use .d.ts for ESM-only bip32, fix trezor type import

Sukuna0007Abhi commented:
> > So, should I revert the changes I did for bip32 & make sure package.json has the right minimum node version
> 
> Yes @Sukuna0007Abhi I think that would be great. thank you!

Pls review

Sukuna0007Abhi committed:
fix: prettier formatting in HermitSignatureImporter

bucko13 commented on packages/caravan-clients/package.json:
```diff
@@ -2,15 +2,29 @@
   "name": "@caravan/clients",
   "version": "1.0.2",
   "description": "A package for querying different bitcoin blockchain backends",
+  "type": "commonjs",
```
would things break if we did this as a module export rather than a commonjs?

bucko13 commented on packages/caravan-psbt/package.json:
```diff
@@ -2,19 +2,26 @@
   "name": "@caravan/psbt",
   "version": "2.0.7",
   "description": "typescript library for working with PSBTs",
+  "type": "commonjs",
```
same here. wondering if this could be module. @Shadouts

bucko13 reviewed (changes requested):
Left some comments and questions. I'm wondering if we should be moving to just the modern module type for all packages. Some other questions about types as well to see if we can avoid `any` especially in newer packages that should be well typed. 

Thanks for bearing with me on this!

Commit:
Merge branch 'main' into issue-262-publint-audit

Commit:
Merge origin/main into issue-262-publint-audit, resolve package.json repository.url conflicts

Made-with: Cursor

Commit:
fix: use main repository.url format to resolve merge conflicts with main

Made-with: Cursor

Commit:
Merge fix-repo-url-only: resolve repository.url conflict with main

Made-with: Cursor

bucko13 commented:
have some conflicts still to resolve and some outstanding questions on types

Commit:
fix: align caravan-bitcoin and caravan-wallets package.json with main to resolve merge conflicts

- Use main's exports/files structure (no .d.mts, flat exports)
- repository.url already matches main
- Ensures clean merge when PR is merged into main

Made-with: Cursor

Legend101Zz commented:
@Sukuna0007Abhi a gentle ping on this , if you had the chance to look into the reviews :)
